HPS/Heartland/Model/Payment.php

```diff
@@ -368,129 +275,629 @@ public function refund(\Magento\Payment\Model\InfoInterface $payment, $amount)
+            }
+
+        }
+        catch (\HpsInvalidRequestException $e) {
+            $errorMsg[] = 'Incorrect parameters on line: ' . $e->getLine() . '. Please get your log files and contact Heartland:
+            ' . $e->getMessage();
+        }
+        catch (\HpsAuthenticationException $e) {
+            $errorMsg[]
+                = 'Authentication on line: ' . $e->getLine() . '. Failure: Credentials Rejected by Gateway please
+                contact Heartland: ' . $e->getMessage();
+        }
+        catch (\HpsGatewayException $e) {
+
+            $errorMsg[] = 'Gateway Error: ' . $e->getMessage();
+
+        }
+        catch (\HpsCreditException $e) {
+            $errorMsg[] = 'Cannot process Payment: ' . $e->getMessage();
+        }
+        catch (\HpsException $e) {
+            $errorMsg[]
+                = 'General Error on line: ' . $e->getLine() . '. The problem will require troubleshooting: ' . $e->getMessage();
+        }
+        catch (\Exception $e) {
+
+            $errorMsg[]
+                = $e->getMessage();
+        }
+        finally { // trying to prevent Magento2 from incorrectly finishing a transaction that has an error
+            // send any error messages from processing to the browser
+            if (count($errorMsg) || ! property_exists($response, 'transactionId') || ! ($response->transactionId > 0
+                ) ) {
+
+                        $errorMsg[]
+                            = 'Please contact this retailer to complete your transaction';
+                //throw new LocalizedException(new Phrase(print_r($errorMsg,true) . " Your transaction could not be
+                //completed!"));
+            }
+
+
+            if (count($errorMsg) && property_exists($response, 'transactionId') && ($response->transactionId > 0
+                ) ) {
+
+                if (($paymentAction === \HpsTransactionType::CHARGE
+                     || $paymentAction === \HpsTransactionType::AUTHORIZE)
+                    && ($response->transactionId > 0)
+                ) {
+                    //Reverse any auth
+                    try{
+
+                        $chargeService = $this->getHpsCreditService();
+                        $chargeService->reverse($response->transactionId,
+                                                $requestedAmount,
+                                                $currency);
+                        unset($successMsg);
+                        $successMsg[] = 'Your transaction was reversed and will not be charged.';
+                    }catch (\Exception $e) {
+                        $errorMsg[]
+                            = $e->getMessage();
+                        $errorMsg[]
+                            = 'Please contact this retailer to complete your transaction';   }
+                }
+                //throw new LocalizedException(new Phrase(print_r($errorMsg,true) . " Your transaction could not be
+                //completed!"));
+            }
```

Comment:
error messages will be added in $errorMsg variable, in case of any transaction errors. This if condition is used to do the reversal process if transaction is completed with an error message